QS Library: Adding The Spin Chain Model

Vanshaj0429 opened the pull request:
## Add Linear Spin Chain Model  

### Description  
This PR adds a **Linear Spin Chain model** to the new `quantum_systems` module.  
The model is implemented as a function in `linear_spin_chain.py` that builds the Hamiltonian, operators, collapse and latex representation for a nearest-neighbor spin chain with configurable length, coupling, and boundary conditions.  

### Implementation details  
- New file: `qutip/quantum_systems/linear_spin_chain.py`  
- Function-based implementation of the linear spin chain Hamiltonian  
- Integrates with the `QuantumSystem` framework introduced in PR #2708  
- Ready for test coverage and further extensions (open vs periodic chains, anisotropic couplings, etc.)  

### Dependency on PR #2708  
⚠️ **Important:** This PR depends on [#2708](https://github.com/qutip/qutip/pull/2708).  
That PR introduces the `QuantumSystem` base class and updates `__init__.py`, which this spin chain function relies on.  

Please merge #2708 first — once it is merged, this PR will only contain the new `linear_spin_chain.py` additions.  

### Next steps  
- Add tests under `qutip/tests/quantum_systems/test_linear_spin_chain.py`

Vanshaj0429 committed:
QS: add LinearSpinChain model

Vanshaj0429 committed:
minor correction in how the system is returned

nwlambert commented on qutip/quantum_systems/linear_spin_chain.py:
```diff
@@ -0,0 +1,311 @@
+        raise ValueError("Chain length N must be at least 2")
+    if N > 15:
+        print(
+            f"Warning: Large system size N={N}. Hilbert space dimension = 2^{N} = {2**N}"
```
i think this warning isn't necessary

Vanshaj0429 committed:
minor corrections in core function

Vanshaj0429 committed:
Add time dependence

Vanshaj0429 committed:
Merge remote-tracking branch 'upstream/dev.qsystem' into linear-spin-chain

Vanshaj0429 committed:
Add spin chain test & init update

Vanshaj0429 commented:
@nwlambert, I have put the tests, can you review the last two commits.

nwlambert commented on qutip/quantum_systems/linear_spin_chain.py:
```diff
@@ -0,0 +1,339 @@
+from typing import Union
+from .quantum_system import QuantumSystem
+
+def _create_sqrt_coefficient(rate):
```
since the same function is used in jaynes_cummings, could import from there?

nwlambert commented on qutip/quantum_systems/linear_spin_chain.py:
```diff
@@ -0,0 +1,339 @@
+        External magnetic field components
+    gamma_dephasing : float or coefficient, default=0.0
+        Pure dephasing rate (T2* process)
+    gamma_depolarizing : float or coefficient, default=0.0
```
an explicit latex description of all the noise channels would be useful (dephasing, polarizing, and thermal)

nwlambert commented on qutip/quantum_systems/linear_spin_chain.py:
```diff
@@ -0,0 +1,339 @@
+    return c_ops
+
+
+def _generate_latex(
```
could we also print the noise lindblads here too? (this case is a bit more complex than JC, with dephasing/depolarizing, so printing what is used would be cool)

Vanshaj0429 replied:
<img width="1087" height="417" alt="Screenshot 2025-09-22 at 7 09 52 PM" src="https://github.com/user-attachments/assets/a700990c-623c-41e7-9e46-de6ba6878f3d" />

I can do something like this

nwlambert commented on qutip/quantum_systems/linear_spin_chain.py:
```diff
@@ -0,0 +1,339 @@
+                beta = 1.0 / temperature
+                exp_factor = np.exp(-beta * transition_frequency)
+                p_down = 1.0 / (1.0 + exp_factor)
+                p_up = exp_factor / (1.0 + exp_factor)
```
just to note, these are the statistics (and hence emission/absorption rates) for a spin-bath? This should also be explicitly mentioned in the doc strings/documentation.

nwlambert replied:
just to clarify a bit, for a spin-system we can imagine it couples to a bath of spins or a bath of bosons.  the choice leads to slightly different emission/absorption rates since bosons follow bose-einstein statistics, but spin bath will follow fermi-direc statistics, which is what you do here.  ok just to support the latter here, for simplicity, but just needs to be documented.

nwlambert commented:
The tests look ok to me.  the final one testing the coefficients could be made a bit more useful?  But am not sure how!  (just fyi, The failing tests are not to do with this PR I think, seems to be a coveralls issue)

Vanshaj0429 committed:
changes in docstring & support lindblad printing

Vanshaj0429 committed:
Add towncrier entry

Vanshaj0429 committed:
more description in docstring, moved helper function to quantum_system

Vanshaj0429 commented:
@Ericgig can you review this please, I have added more detail in the docstring to make it more readable and moved the `_create_sqrt_coefficient` function to `quantum_system.py`. I think these were two prominent issues.

Ericgig reviewed (commented):
I don't think `linear_spin_chain()` should not be valid without any arguments.
At least let's make `N`, `model` and `J` required...

Vanshaj0429 commented:
> @Vanshaj0429 thank you for your work! Please let us know whether you have an intention and time address the rest of the code review. Since this PR has been inactive for more than 6 months, we will have to close it otherwise.

Hi @veronikakurth, apologies for not responding sooner and for the prolonged inactivity on this PR.

In early November I came down with a respiratory infection that escalated into a serious illness, and I had to stop all work and take leave for medical treatment. I only started working again in late February, and even then had other pressing commitments to catch up on first. The situation unfortunately cost me a number of opportunities, and this PR was one of the things I wasn't able to get back to as quickly as I'd have liked.

Thank you for taking the time to flag this rather than simply closing it. I want to see this work through to completion and am happy to address the remaining review comments and should be able to hopefuly complete this in a week's time.
